ethers-core/ethers-derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+//! )]
+//! pub struct Puzzle {
+//!     pub organization: H160,
+//!     pub contributor: H160,
+//!     pub commit: String,
+//!     pub project: String,
+//! }
+//!
+//! let puzzle = Puzzle {
+//!     organization: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     contributor: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     commit: "5693b7019eb3e4487a81273c6f5e1832d77acb53".to_string(),
+//!     project: "radicle-reward".to_string(),
+//! };
+//!
+//! let hash = puzzle.encode_eip712()?;
+//!
+//! let decoded: Vec<u8> = hex::decode(hash).expect("failed to decode")
+//! let byte_array: [u8; 32] = <[u8; 32]>::try_from(&decoded[..])?;
+//! ```
+//!
+//! # Limitations
+//!
+//! At the moment, the derive macro does not recursively encode nested Eip712 structs.
+//!
+
+use proc_macro::TokenStream;
+use quote::quote;
+
+// import eip712 utilities from ethers_core::types::transaction::eip712
+use ethers_core::types::transaction::eip712;
+
+// Main implementation macro, used to compute static values and define
+// method for encoding the final eip712 payload;
+fn impl_eip_712_macro(ast: &syn::DeriveInput) -> TokenStream {
+    // Primary type should match the type in the ethereum verifying contract;
+    let primary_type = &ast.ident;
+
+    // Computer domain separator
+    let domain_attributes: eip712::Eip712Domain = eip712::Eip712Domain::from(ast);
+    let domain_separator = domain_attributes.separator();
+    let domain_type_hash = hex::encode(eip712::eip712_domain_type_hash());
+
+    // Must parse the AST at compile time.
+    let parsed_fields = eip712::parse_fields(ast);
+
+    // JSON Stringify the field names and types to pass into the
+    // derived encode_eip712() method as a static string;
+    // the AST of the struct is not available at runtime, so this is
+    // a work around for passing in the struct fields;
+    let fields: String = serde_json::to_string(&parsed_fields)
+        .expect("failed to serialize parsed fields into JSON string");
+
+    // Compute the type hash for the derived struct using the parsed fields from above;
+    let type_hash = eip712::make_type_hash(primary_type.clone().to_string(), &parsed_fields);
+
+    let implementation = quote! {
+        impl Eip712 for #primary_type {
+            type Error = ethers_core::types::transaction::eip712::Eip712Error;
+
+            fn type_hash() -> String {
+                #type_hash.to_string()
+            }
+
+            fn domain_separator() -> String {
+                #domain_separator.to_string()
+            }
+
+            fn encode_eip712(&self) -> Result<String, Self::Error> {
+                // Ok(make_struct_hash(self, #domain_separator, #type_hash, #fields)?.to_string())
+                let json: serde_json::Value = serde_json::from_str(#fields)?;
```

Comment:
this is a clever hack lol
this looks rather similar to the `EthEvent` implementation.
I might be wrong but all this does is essentially encode the whole struct (its fields) into a `Vec<Token>` right?
can't we mandate that this struct implements `Tokenizable` which provides `Tokenizable::into_token() ->Token` which is a tuple for structs. This is all implemented automatically with `EthAbiType` macro, examples https://github.com/gakonst/ethers-rs/blob/master/ethers-contract/tests/common/derive.rs
This will require a clone however, we could get rid of this with some effort but I guess cloning would be negligeble.

Reply:
Oh nice! `EthAbiType` looks great! Will look for solutions to avoid needing to serialize the hash map here.

Reply:
No longer using JSON to pass values around. Using `EthAbiType` for tokenizing the struct. Thanks for the tip!